relay/writer/file_util.go

```diff
@@ -0,0 +1,268 @@
+			syncStartPos, ev2.Header, ev.Header)
+	}
+
+	// duplicate in the file
+	return true, nil
+}
+
+// getTxnPosGTIDs gets position/GTID set for all completed transactions from a binlog file.
+// NOTE: we use a int64 rather than a uint32 to represent the latest transaction's end log pos.
+func getTxnPosGTIDs(filename string, p *parser.Parser) (int64, gtid.Set, error) {
+	// use a FileReader to parse the binlog file.
+	rCfg := &reader.FileReaderConfig{
+		EnableRawMode: false, // in order to get GTID set, we always disable RawMode.
+	}
+	startPos := gmysql.Position{Name: filename, Pos: 0} // always start from the file header
+	r := reader.NewFileReader(rCfg)
+	defer r.Close()
+	err := r.StartSyncByPos(startPos) // we always parse the file by pos
+	if err != nil {
+		return 0, nil, errors.Annotatef(err, "start sync by pos %s for %s", startPos, filename)
+	}
+
+	var (
+		latestPos   int64
+		latestGSet  gmysql.GTIDSet
+		nextGTIDStr string // can be recorded if the coming transaction completed
+		flavor      string
+	)
+	for {
+		var e *replication.BinlogEvent
+		ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+		e, err = r.GetEvent(ctx)
+		cancel()
+		if err != nil {
+			break // now, we stop to parse for any errors
+		}
+
+		// NOTE: only update pos/GTID set for DDL/XID to get an complete transaction.
+		switch ev := e.Event.(type) {
+		case *replication.QueryEvent:
+			isDDL := common.CheckIsDDL(string(ev.Query), p)
+			if isDDL {
+				if latestGSet != nil { // GTID may not be enabled in the binlog
+					err = latestGSet.Update(nextGTIDStr)
+					if err != nil {
+						return 0, nil, errors.Annotatef(err, "update GTID set %v with GTID %s", latestGSet, nextGTIDStr)
+					}
+				}
+				latestPos = int64(e.Header.LogPos)
+			}
+		case *replication.XIDEvent:
+			if latestGSet != nil { // GTID may not be enabled in the binlog
+				err = latestGSet.Update(nextGTIDStr)
+				if err != nil {
+					return 0, nil, errors.Annotatef(err, "update GTID set %v with GTID %s", latestGSet, nextGTIDStr)
+				}
+			}
+			latestPos = int64(e.Header.LogPos)
+		case *replication.GTIDEvent:
+			if latestGSet == nil {
+				return 0, nil, errors.Errorf("should have a PreviousGTIDsEvent before the GTIDEvent %+v", e.Header)
+			}
+			// learn from: https://github.com/siddontang/go-mysql/blob/c6ab05a85eb86dc51a27ceed6d2f366a32874a24/replication/binlogsyncer.go#L736
+			u, _ := uuid.FromBytes(ev.SID)
+			nextGTIDStr = fmt.Sprintf("%s:%d", u.String(), ev.GNO)
+		case *replication.MariadbGTIDEvent:
+			if latestGSet == nil {
+				return 0, nil, errors.Errorf("should have a MariadbGTIDListEvent before the MariadbGTIDEvent %+v", e.Header)
+			}
+			// learn from: https://github.com/siddontang/go-mysql/blob/c6ab05a85eb86dc51a27ceed6d2f366a32874a24/replication/binlogsyncer.go#L745
+			GTID := ev.GTID
+			nextGTIDStr = fmt.Sprintf("%d-%d-%d", GTID.DomainID, GTID.ServerID, GTID.SequenceNumber)
+		case *replication.MariadbGTIDListEvent:
```

Comment:
I remember  event parser of `go-mysql` handles these logic, can we use it's logic in file reader?

Reply:
you mean computes GTID set in `FileReader` rather than in this _file_util_?
I want `FileReader` returns uniform binlog event to the caller, but the `replication.BinlogEvent` cannot carry other fields easily.